utilities/console.py

```
@@ -107,16 +115,40 @@ def disconnect(self):
+
+        Uses pty.openpty() + subprocess.Popen instead of pexpect.spawn to avoid
+        os.forkpty(), which is deprecated in multi-threaded processes (Python 3.12+).
+        """
+        master_fd, slave_fd = pty.openpty()
+        self._proc = subprocess.Popen(
+            shlex.split(self.cmd),
+            stdin=slave_fd,
+            stdout=slave_fd,
+            stderr=slave_fd,
+            start_new_session=True,
+            preexec_fn=lambda: fcntl.ioctl(0, termios.TIOCSCTTY, 0),
```

Ruff (0.15.14)

[warning] `preexec_fn` argument is unsafe when using threads

(PLW1509)